File: python/examples/symcache.py
#!/usr/bin/env python3
"""symcache — v0.9.0 Phase 1 symbol-cache validator.

Output mirrors c/examples/symcache.c + go/cmd/symcache (modulo dt=
timings).

SPDX-License-Identifier: MIT
"""
import argparse
import sys
import time

import bpclient


def now_ms() -> float:
    return time.monotonic() * 1000.0


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--path", default="P:1,S:2")
    ap.add_argument("--tag", default="OCX_TEST")
    args = ap.parse_args()

    try:
        c = bpclient.Client()
        c.open()
    except Exception as e:
        print(f"[symcache] open failed: {e}", file=sys.stderr)
        return 2
    try:
        c.open_session()
        db = c.open_tagdb(args.path)
        try:
            t0 = now_ms()
            n = db.build()
            t1 = now_ms()
            print(f"[symcache] path={args.path} build n={n} dt={t1 - t0:.2f}ms")

            l0 = now_ms()
            info = db.lookup_symbol(args.tag)
            l1 = now_ms()
            print(f"[symcache] lookup#1 cold dt={l1 - l0:.2f}ms  "
                  f"data_type=0x{info.data_type:04x} elem_byte_size={info.elem_byte_size}")

            l2 = now_ms()
            db.lookup_symbol(args.tag)
            l3 = now_ms()
            print(f"[symcache] lookup#2 warm dt={l3 - l2:.3f}ms  (cache hit)")

            p0 = now_ms()
            try:
                db.preload_symbols()
                rc = 0
            except Exception as e:
                rc = bpclient.err_code(e)
            p1 = now_ms()
            print(f"[symcache] preload all dt={p1 - p0:.2f}ms rc={rc}")

            l4 = now_ms()
            db.lookup_symbol(args.tag)
            l5 = now_ms()
            print(f"[symcache] lookup#3 after-preload dt={l5 - l4:.3f}ms  rc=0")
            print("[symcache] PASS")
            return 0
        finally:
            db.close()
    finally:
        c.close()


if __name__ == "__main__":
    sys.exit(main())

File: c/examples/symcache.c
/*
 * symcache.c — v0.9.0 Phase 1 symbol-cache validator.
 *
 * Opens a tag DB for the L85 (slot 2 by default), builds it, then:
 *   - Calls bp_tagdb_lookup_symbol("OCX_TEST") TWICE and reports the
 *     wall-clock latency of each.  First call walks the PLC's table
 *     incrementally and caches every examined symbol; second call is
 *     an in-memory hit and should be sub-millisecond.
 *   - Calls bp_tagdb_preload_symbols and prints its timing.
 *
 * Output is byte-identical (modulo dt= numbers) across the three
 * SDKs; the cross-language gate is the diff of the two summary lines.
 *
 * SPDX-License-Identifier: MIT
 */
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "bpclient.h"

static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1.0e6;
}

int main(int argc, char *argv[]) {
    const char *path = "P:1,S:2";
    const char *tag  = "OCX_TEST";

    static struct option opts[] = {
        {"path", required_argument, 0, 'p'},
        {"tag",  required_argument, 0, 't'},
        {0,0,0,0}
    };
    int oc, idx;
    while ((oc = getopt_long(argc, argv, "p:t:", opts, &idx)) != -1) {
        if      (oc == 'p') path = optarg;
        else if (oc == 't') tag  = optarg;
    }

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "[symcache] open failed\n");
        return 2;
    }
    bp_client_open_session(cl, NULL);

    bp_tagdb_t *db = NULL;
    int rc = bp_tagdb_open(cl, path, &db);
    if (rc != BP_OK) {
        fprintf(stderr, "[symcache] tagdb_open rc=%d (%s)\n", rc, bp_strerror(rc));
        bp_client_close(cl);
        return 1;
    }
    uint16_t n = 0;
    double t0 = now_ms();
    rc = bp_tagdb_build(db, &n);
    double t1 = now_ms();
    if (rc != BP_OK) {
        fprintf(stderr, "[symcache] build rc=%d\n", rc);
        bp_tagdb_close(db); bp_client_close(cl); return 1;
    }
    printf("[symcache] path=%s build n=%u dt=%.2fms\n", path, n, t1 - t0);

    bp_symbol_info_t info;
    /* First lookup — walks the cache lazily. */
    double l0 = now_ms();
    rc = bp_tagdb_lookup_symbol(db, tag, &info);
    double l1 = now_ms();
    if (rc != BP_OK) {
        fprintf(stderr, "[symcache] lookup#1 rc=%d (%s)\n", rc, bp_strerror(rc));
        bp_tagdb_close(db); bp_client_close(cl); return 1;
    }
    printf("[symcache] lookup#1 cold dt=%.2fms  data_type=0x%04x elem_byte_size=%u\n",
           l1 - l0, info.data_type, info.elem_byte_size);

    /* Second lookup — should be a cached hit. */
    double l2 = now_ms();
    rc = bp_tagdb_lookup_symbol(db, tag, &info);
    double l3 = now_ms();
    if (rc != BP_OK) {
        fprintf(stderr, "[symcache] lookup#2 rc=%d\n", rc);
        bp_tagdb_close(db); bp_client_close(cl); return 1;
    }
    printf("[symcache] lookup#2 warm dt=%.3fms  (cache hit)\n", l3 - l2);

    /* Eager preload of the rest. */
    double p0 = now_ms();
    rc = bp_tagdb_preload_symbols(db);
    double p1 = now_ms();
    printf("[symcache] preload all dt=%.2fms rc=%d\n", p1 - p0, rc);

    /* Now any name fetched after preload is a cache hit. */
    double l4 = now_ms();
    rc = bp_tagdb_lookup_symbol(db, tag, &info);
    double l5 = now_ms();
    printf("[symcache] lookup#3 after-preload dt=%.3fms  rc=%d\n", l5 - l4, rc);

    printf("[symcache] PASS\n");
    bp_tagdb_close(db);
    bp_client_close(cl);
    return 0;
}
